gqlalchemy/transformations/translators/dgl_translator.py

```diff
@@ -0,0 +1,137 @@
+        host: str = MG_HOST,
+        port: int = MG_PORT,
+        username: str = MG_USERNAME,
+        password: str = MG_PASSWORD,
+        encrypted: bool = MG_ENCRYPTED,
+        client_name: str = MG_CLIENT_NAME,
+        lazy: bool = MG_LAZY,) -> None:
+        super().__init__(default_node_label, default_edge_type, host, port, username, password, encrypted, client_name, lazy)
+
+    def to_cypher_queries(self, graph):
+        """Produce cypher queries for data saved as part of the DGL graph. The method handles both homogeneous and heterogeneous graph. If the graph is homogeneous, a default DGL's labels will be used.
+         _N as a node label and _E as edge label. The method converts 1D as well as multidimensional features. If there are some isolated nodes inside DGL graph, they won't get transferred. Nodes and edges
+         created in Memgraph DB will, for the consistency reasons, have property `dgl_id` set to the id they have as part of the DGL graph. Note that this method doesn't insert anything inside the database,
+         it just creates cypher queries. To insert queries the following code can be used:
+         >>> memgraph = Memgraph()
+         dgl_graph = DGLGraph(...)
+         for query in DGLTranslator().to_cypher_queries(dgl_graph):
+            memgraph.execute(query)
+
+         Args:
+            graph: A reference to the DGL graph.
+        Returns:
+            cypher queries.
+        """
+        queries = []
+        if graph is None:
+            return queries
+        for etype in graph.canonical_etypes:
+            source_node_label, edge_type, dest_node_label = etype
+            source_nodes, dest_nodes, eids = graph.edges(etype=etype, form="all")
+            # Get label and type properties
+            node_src_label_properties = graph.nodes[source_node_label].data
+            node_dest_label_properties = graph.nodes[dest_node_label].data
+            etype_properties = graph.edges[etype].data
+            for source_node_id, dest_node_id, eid in zip(source_nodes, dest_nodes, eids):
+                # Handle properties
+                source_node_properties, dest_node_properties, edge_properties = {}, {}, {}
+                # Copy source node properties
+                source_node_properties = dict(map(lambda pair: (pair[0], to_cypher_value(pair[1][source_node_id])), node_src_label_properties.items()))
+                source_node_properties[DGL_ID] = int(source_node_id)
+                # Copy destination node properties
+                dest_node_properties = dict(map(lambda pair: (pair[0], to_cypher_value(pair[1][dest_node_id])), node_dest_label_properties.items()))
+                dest_node_properties[DGL_ID] = int(dest_node_id)
+                # Copy edge features
+                edge_properties = dict(map(lambda pair: (pair[0], to_cypher_value(pair[1][eid])), etype_properties.items()))
+                edge_properties[DGL_ID] = int(eid)
```

Comment:
If I understand this correctly, you are going over source_nodes, dest_nodes and eids. And for all those pairs, node_src_label_properties which I persume represents properties keys, is same for all source nodes? Why is that?

Couldn't it happen that some of nodes in source nodes have extra property? Or this is not stored at all?

Reply:
Yes so `node_src_label_properties` and `node_dest_label_properties` are used in `DGL` for storing properties. They save for each node type an array, where each element of the array contains data about one node. `DGL` stores only numeric properties and I have imposed the restriction that only properties which are on all nodes, can be translated to `DGL` and `PyG` because both frameworks don't allow different situation.